core/server/src/test/java/alluxio/worker/netty/DataServerBlockReadHandlerTest.java

```diff
@@ -0,0 +1,304 @@
+    mChannel = new EmbeddedChannel(
+        new DataServerBlockReadHandler(NettyExecutors.BLOCK_READER_EXECUTOR, mBlockWorker,
+            FileTransferType.TRANSFER));
+
+    long fileSize = PACKET_SIZE * 2;
+    populateInputFile(fileSize, 0, fileSize - 1);
+
+    BlockReader blockReader = PowerMockito.spy(mBlockReader);
+    // Do not call close here so that we can check result. It will be closed explicitly.
+    PowerMockito.doNothing().when(blockReader).close();
+    PowerMockito
+        .when(mBlockWorker.readBlockRemote(Mockito.anyLong(), Mockito.anyLong(), Mockito.anyLong()))
+        .thenReturn(blockReader);
+    mChannel.writeInbound(buildReadRequest(0, fileSize));
+    checkAllReadResponses();
+    mBlockReader.close();
+  }
+
+  @Test
+  public void readEmptyFile() throws Exception {
+    populateInputFile(0, 0, 0);
+    mChannel.writeInbound(buildReadRequest(0, 0));
+    Object response = waitForOneResponse();
+    checkReadResponse(response, Protocol.Status.Code.INVALID_ARGUMENT);
+    waitForChannelClose();
+    Assert.assertTrue(!mChannel.isOpen());
+  }
+
+  @Test
+  public void cancelRequest() throws Exception {
+    long fileSize = PACKET_SIZE * 10 + 1;
+    populateInputFile(fileSize, 0, fileSize - 1);
+    RPCProtoMessage readRequest = buildReadRequest(0, fileSize);
+    Protocol.ReadRequest request = (Protocol.ReadRequest) readRequest.getMessage();
+    RPCProtoMessage cancelRequest =
+        new RPCProtoMessage(request.toBuilder().setCancel(true).build(), null);
+    mChannel.writeInbound(readRequest);
+    mChannel.writeInbound(cancelRequest);
+
+    // Make sure we can still get EOF after cancelling though the read request is not necessarily
+    // fulfilled.
+    boolean eof = false;
+    long maxIterations = 100;
+    while (maxIterations > 0) {
+      Object response = waitForOneResponse();
+      DataBuffer buffer = checkReadResponse(response, Protocol.Status.Code.OK);
+      if (buffer == null) {
+        eof = true;
+        break;
+      }
+      buffer.release();
+      maxIterations--;
+      Assert.assertTrue(mChannel.isOpen());
+    }
+    Assert.assertTrue(eof);
+  }
+
+  @Test
+  public void readFailure() throws Exception {
+    long fileSize = PACKET_SIZE * 10 + 1;
+    populateInputFile(0, 0, fileSize - 1);
+    mBlockReader.close();
+    mChannel.writeInbound(buildReadRequest(0, fileSize));
+    Object response = waitForOneResponse();
+    checkReadResponse(response, Protocol.Status.Code.INTERNAL);
+    waitForChannelClose();
+    Assert.assertTrue(!mChannel.isOpen());
+  }
+
+  /**
+   * Populates the input file, also computes the checksum for part of the file.
+   *
+   * @param length the length of the file
+   * @param start the start position to compute the checksum
+   * @param end the last position to compute the checksum
+   * @throws Exception if it fails to populate the input file
+   */
+  private void populateInputFile(long length, long start, long end) throws Exception {
```

Comment:
Is there a reason why the checksum is not returned, but instead modifies a member variable?